<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRECAST: simulation • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRECAST: simulation">
<meta property="og:description" content="PRECAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRECAST: simulation</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2022-06-07</h4>
      
      
      <div class="hidden name"><code>PRECAST.Simu.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="generate-the-simulated-data">Generate the simulated data<a class="anchor" aria-label="anchor" href="#generate-the-simulated-data"></a>
</h2>
<p>First, we generate the multiple spatial transcriptomics data with
lattice neighborhood, i.e. ST platform by using the function
<code>gendata_seulist</code> in <code>PRECAST</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PRECAST</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gendata_seulist.html">gendata_seulist</a></span><span class="op">(</span>height1<span class="op">=</span><span class="fl">20</span>, width1<span class="op">=</span><span class="fl">20</span>, height2<span class="op">=</span><span class="fl">24</span>, width2<span class="op">=</span><span class="fl">25</span>,p<span class="op">=</span><span class="fl">200</span>, K<span class="op">=</span><span class="fl">4</span>, alpha<span class="op">=</span><span class="fl">20</span>,     sigma2<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>
<span class="va">seuList</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 200 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (200 features, 0 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 200 features across 600 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (200 features, 0 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> 
<span class="co">#&gt;          orig.ident nCount_RNA nFeature_RNA row col true_cluster</span>
<span class="co">#&gt; cell1_1       cell1       2957          118   1   1            2</span>
<span class="co">#&gt; cell1_2       cell1       2005          112   2   1            3</span>
<span class="co">#&gt; cell1_3       cell1       3680          117   3   1            1</span>
<span class="co">#&gt; cell1_4       cell1       3070          113   4   1            2</span>
<span class="co">#&gt; cell1_5       cell1       3419          112   5   1            2</span>
<span class="co">#&gt; cell1_6       cell1       4324          122   6   1            1</span>
<span class="co">#&gt; cell1_7       cell1       4239          117   7   1            1</span>
<span class="co">#&gt; cell1_8       cell1       5127          112   8   1            1</span>
<span class="co">#&gt; cell1_9       cell1       4584          113   9   1            1</span>
<span class="co">#&gt; cell1_10      cell1       3452          121  10   1            1</span>
<span class="co">## Must include the columns named "row" and "col" for saving the spatial coordinates</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-precast-using-simulated-data">Fit PRECAST using simulated data<a class="anchor" aria-label="anchor" href="#fit-precast-using-simulated-data"></a>
</h2>
<div class="section level3">
<h3 id="prepare-the-precastobject-with-preprocessing-step-">Prepare the PRECASTObject with preprocessing step.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-with-preprocessing-step-"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Create </span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">)</span>
<span class="co">#&gt; Filter spots and features from Raw count data...</span>
<span class="co">#&gt;  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Find the spatially variables genes by SPARK-X...</span>
<span class="co">#&gt; ## ===== SPARK-X INPUT INFORMATION ====</span>
<span class="co">#&gt; ## number of total samples: 400</span>
<span class="co">#&gt; ## number of total genes: 189</span>
<span class="co">#&gt; ## Running with single core, may take some time</span>
<span class="co">#&gt; ## Testing With Projection Kernel</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 1</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 2</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 3</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 4</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 5</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 1</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 2</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 3</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 4</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 5</span>
<span class="co">#&gt; Find the spatially variables genes by SPARK-X...</span>
<span class="co">#&gt; ## ===== SPARK-X INPUT INFORMATION ====</span>
<span class="co">#&gt; ## number of total samples: 600</span>
<span class="co">#&gt; ## number of total genes: 192</span>
<span class="co">#&gt; ## Running with single core, may take some time</span>
<span class="co">#&gt; ## Testing With Projection Kernel</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 1</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 2</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 3</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 4</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 5</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 1</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 2</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 3</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 4</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 5</span>
<span class="co">#&gt; Warning in FUN(X[[i]], ...): Only 189 SVGs will be returned since the number of genes is less than 2000</span>
<span class="co">#&gt; Warning in FUN(X[[i]], ...): Only 192 SVGs will be returned since the number of genes is less than 2000</span>
<span class="co">#&gt; Warning in CreatePRECASTObject(seuList): Number of SVGs in one of sample is less</span>
<span class="co">#&gt; than 2000, so set minimum number of SVGs as gene.number=189</span>
<span class="co">#&gt; Select common SVGs  for multiple samples...</span>
<span class="co">#&gt; Filter spots and features from SVGs count data...</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## check the number of genes/features after filtering step</span>
<span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 183 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (183 features, 183 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 183 features across 600 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (183 features, 183 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>

<span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/AddAdjList.html">AddAdjList</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, platform <span class="op">=</span> <span class="st">"ST"</span><span class="op">)</span>
<span class="co">#&gt; Neighbors were identified for 400 out of 400 spots.</span>
<span class="co">#&gt; Neighbors were identified for 600 out of 600 spots.</span>

<span class="co">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the information in the algorithm.</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal<span class="op">=</span><span class="cn">TRUE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, seed<span class="op">=</span><span class="fl">2022</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-precast">Fit PRECAST<a class="anchor" aria-label="anchor" href="#fit-precast"></a>
</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters <span class="math inline">\(K\)</span> or set <code>K</code> to
be an integer vector by using modified BIC(MBIC) to determine <span class="math inline">\(K\)</span>. First, we try using user-specified
number of clusters. For convenience, we give a single K here.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Given K</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj, <span class="at">K=</span><span class="dv">5</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Intergrative data info.: 2 samples, 183 genes X 1000 spots------</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; PRECAST model setting: error_heter=TRUE, Sigma_equal=TRUE, Sigma_diag=TRUE, mix_prop_heter=TRUE</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Start computing intial values...</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; fitting ...</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="er">|</span>                                                                      <span class="er">|</span>   <span class="dv">0</span>%</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="er">|===================================</span>                                   <span class="er">|</span>  <span class="dv">50</span>%</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="sc">|</span>                                                                            </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="er">|======================================================================|</span> <span class="dv">100</span>%</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; ----Fitting PRECAST model----------------</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; variable initialize finish! </span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 2, loglik= 17763.335938, dloglik=1.000008 </span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; diff Energy = 4.481528 </span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; diff Energy = 1.010073 </span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 3, loglik= 22625.228516, dloglik=0.273704 </span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; diff Energy = 0.628834 </span></span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 4, loglik= 24101.154297, dloglik=0.065234 </span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 5, loglik= 24891.617188, dloglik=0.032798 </span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 6, loglik= 25366.353516, dloglik=0.019072 </span></span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 7, loglik= 25646.306641, dloglik=0.011036 </span></span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 8, loglik= 25830.437500, dloglik=0.007180 </span></span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 9, loglik= 25959.492188, dloglik=0.004996 </span></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 10, loglik= 26054.701172, dloglik=0.003668 </span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 11, loglik= 26127.556641, dloglik=0.002796 </span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 12, loglik= 26184.681641, dloglik=0.002186 </span></span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 13, loglik= 26230.121094, dloglik=0.001735 </span></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 14, loglik= 26266.546875, dloglik=0.001389 </span></span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 15, loglik= 26295.830078, dloglik=0.001115 </span></span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 16, loglik= 26319.435547, dloglik=0.000898 </span></span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 17, loglik= 26338.500000, dloglik=0.000724 </span></span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 18, loglik= 26353.927734, dloglik=0.000586 </span></span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 19, loglik= 26366.476562, dloglik=0.000476 </span></span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; predict Y and V! </span></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; Finish ICM step! </span></span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; iter = 20, loglik= 26376.701172, dloglik=0.000388</span></span></code></pre></div>
<p>Select a best model and use ARI to check the performance of
clustering</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## backup the fitting results in resList</span>
<span class="va">resList</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span>
<span class="co"># PRECASTObj@resList &lt;- resList</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">selectModel</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span>
<span class="co">#&gt; Sigma is set to diagonal matrices</span>
<span class="va">true_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">seuList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">true_cluster</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ : Named num [1:400] 2 3 1 2 2 1 1 1 1 1 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:400] "cell1_1" "cell1_2" "cell1_3" "cell1_4" ...</span>
<span class="co">#&gt;  $ : Named num [1:600] 3 3 3 1 1 1 1 1 1 1 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:600] "cell2_1" "cell2_2" "cell2_3" "cell2_4" ...</span>
<span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.9847307</span></code></pre></div>
<p>Integrate the two samples by the function
<code>IntegrateSpaData</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species<span class="op">=</span><span class="st">'unknown'</span><span class="op">)</span>
<span class="va">seuInt</span> 
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 183 features across 1000 samples within 1 assay </span>
<span class="co">#&gt; Active assay: PRE_CAST (183 features, 0 variable features)</span>
<span class="co">#&gt;  2 dimensional reductions calculated: PRECAST, position</span>
<span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>,point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p12</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># users can plot each sample by setting combine=FALSE</span></code></pre></div>
<p>Show the spatial UMAP/tNSE RGB plot</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span> 
<span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>,item<span class="op">=</span><span class="st">'RGB_UMAP'</span>,point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">TRUE</span>, text_size<span class="op">=</span><span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#seuInt &lt;- AddTSNE(seuInt) </span>
<span class="co">#SpaPlot(seuInt, batch=NULL,item='RGB_TSNE',point_size=2, combine=T, text_size=15)</span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E04D50"</span>, <span class="st">"#4374A5"</span>, <span class="st">"#F08A21"</span>,<span class="st">"#2AB673"</span>, <span class="st">"#FCDDDE"</span>,  <span class="st">"#70B5B0"</span>, <span class="st">"#DFE0EE"</span> ,<span class="st">"#D0B14C"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>,  font_family<span class="op">=</span><span class="st">'serif'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span><span class="op">)</span> <span class="co"># Times New Roman</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item<span class="op">=</span><span class="st">'batch'</span>, point_size <span class="op">=</span> <span class="fl">1</span>,  font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> </code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># It is noted that only sample batch 1 has cluster 4, and only sample batch 2 has cluster 7. </span></code></pre></div>
<p>Show the UMAP plot based on the extracted features from PRECAST.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'UMAP3'</span>, item<span class="op">=</span><span class="st">'cluster'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Users can also use the visualization functions in Seurat package:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'position'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'tSNE'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-2.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'PRECAST'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-3.png" width="700"></p>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Calculating cluster 1</span>
<span class="co">#&gt; Calculating cluster 2</span>
<span class="co">#&gt; Calculating cluster 3</span>
<span class="co">#&gt; Calculating cluster 4</span>
<span class="co">#&gt; Calculating cluster 5</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span>
<span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span>

<span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Centering and scaling data matrix</span>
<span class="va">seus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seuInt</span>, downsample <span class="op">=</span> <span class="fl">400</span><span class="op">)</span>
<span class="va">color_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>

<span class="co">## HeatMap</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doHeatmap.html">doHeatmap</a></span><span class="op">(</span><span class="va">seus</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span>, cell_label<span class="op">=</span> <span class="st">"Domain"</span>,
                grp_label <span class="op">=</span> <span class="cn">F</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>,
                pt_size<span class="op">=</span><span class="fl">6</span>,slot <span class="op">=</span> <span class="st">'scale.data'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">16</span><span class="op">)</span>,
        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">18</span>, face<span class="op">=</span><span class="st">'bold'</span><span class="op">)</span>,
        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">7</span>, face<span class="op">=</span> <span class="st">"italic"</span>, family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22000)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=Chinese (Simplified)_China.936 </span>
<span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.936   </span>
<span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.936</span>
<span class="co">#&gt; [4] LC_NUMERIC=C                              </span>
<span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.936    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.3.5      dplyr_1.0.8        patchwork_1.1.1    PRECAST_1.0       </span>
<span class="co">#&gt; [5] SeuratObject_4.0.4 Seurat_4.1.0      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2                  reticulate_1.24            </span>
<span class="co">#&gt;   [3] tidyselect_1.1.2            htmlwidgets_1.5.4          </span>
<span class="co">#&gt;   [5] grid_4.1.2                  BiocParallel_1.28.3        </span>
<span class="co">#&gt;   [7] Rtsne_0.15                  munsell_0.5.0              </span>
<span class="co">#&gt;   [9] ScaledMatrix_1.2.0          codetools_0.2-18           </span>
<span class="co">#&gt;  [11] ragg_1.2.2                  ica_1.0-2                  </span>
<span class="co">#&gt;  [13] future_1.24.0               miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [15] withr_2.5.0                 spatstat.random_2.1-0      </span>
<span class="co">#&gt;  [17] colorspace_2.0-3            Biobase_2.54.0             </span>
<span class="co">#&gt;  [19] highr_0.9                   knitr_1.37                 </span>
<span class="co">#&gt;  [21] rstudioapi_0.13             stats4_4.1.2               </span>
<span class="co">#&gt;  [23] SingleCellExperiment_1.16.0 ROCR_1.0-11                </span>
<span class="co">#&gt;  [25] tensor_1.5                  listenv_0.8.0              </span>
<span class="co">#&gt;  [27] MatrixGenerics_1.6.0        labeling_0.4.2             </span>
<span class="co">#&gt;  [29] GenomeInfoDbData_1.2.7      polyclip_1.10-0            </span>
<span class="co">#&gt;  [31] farver_2.1.0                rprojroot_2.0.2            </span>
<span class="co">#&gt;  [33] parallelly_1.30.0           vctrs_0.3.8                </span>
<span class="co">#&gt;  [35] generics_0.1.2              xfun_0.29                  </span>
<span class="co">#&gt;  [37] ggthemes_4.2.4              R6_2.5.1                   </span>
<span class="co">#&gt;  [39] GenomeInfoDb_1.30.1         ggbeeswarm_0.6.0           </span>
<span class="co">#&gt;  [41] rsvd_1.0.5                  bitops_1.0-7               </span>
<span class="co">#&gt;  [43] spatstat.utils_2.3-0        cachem_1.0.6               </span>
<span class="co">#&gt;  [45] DelayedArray_0.20.0         assertthat_0.2.1           </span>
<span class="co">#&gt;  [47] promises_1.2.0.1            scales_1.1.1               </span>
<span class="co">#&gt;  [49] beeswarm_0.4.0              gtable_0.3.0               </span>
<span class="co">#&gt;  [51] beachmat_2.10.0             globals_0.14.0             </span>
<span class="co">#&gt;  [53] goftest_1.2-3               rlang_1.0.1                </span>
<span class="co">#&gt;  [55] systemfonts_1.0.4           splines_4.1.2              </span>
<span class="co">#&gt;  [57] lazyeval_0.2.2              spatstat.geom_2.3-2        </span>
<span class="co">#&gt;  [59] yaml_2.3.5                  reshape2_1.4.4             </span>
<span class="co">#&gt;  [61] abind_1.4-5                 httpuv_1.6.5               </span>
<span class="co">#&gt;  [63] tools_4.1.2                 ellipsis_0.3.2             </span>
<span class="co">#&gt;  [65] spatstat.core_2.4-0         jquerylib_0.1.4            </span>
<span class="co">#&gt;  [67] RColorBrewer_1.1-2          BiocGenerics_0.40.0        </span>
<span class="co">#&gt;  [69] ggridges_0.5.3              Rcpp_1.0.8                 </span>
<span class="co">#&gt;  [71] plyr_1.8.6                  sparseMatrixStats_1.6.0    </span>
<span class="co">#&gt;  [73] zlibbioc_1.40.0             purrr_0.3.4                </span>
<span class="co">#&gt;  [75] RCurl_1.98-1.6              rpart_4.1.16               </span>
<span class="co">#&gt;  [77] deldir_1.0-6                pbapply_1.5-0              </span>
<span class="co">#&gt;  [79] viridis_0.6.2               cowplot_1.1.1              </span>
<span class="co">#&gt;  [81] S4Vectors_0.32.3            zoo_1.8-9                  </span>
<span class="co">#&gt;  [83] SummarizedExperiment_1.24.0 ggrepel_0.9.1              </span>
<span class="co">#&gt;  [85] cluster_2.1.2               fs_1.5.2                   </span>
<span class="co">#&gt;  [87] magrittr_2.0.2              GiRaF_1.0.1                </span>
<span class="co">#&gt;  [89] data.table_1.14.2           scattermore_0.8            </span>
<span class="co">#&gt;  [91] lmtest_0.9-39               RANN_2.6.1                 </span>
<span class="co">#&gt;  [93] fitdistrplus_1.1-6          matrixStats_0.61.0         </span>
<span class="co">#&gt;  [95] mime_0.12                   evaluate_0.15              </span>
<span class="co">#&gt;  [97] xtable_1.8-4                mclust_5.4.9               </span>
<span class="co">#&gt;  [99] IRanges_2.28.0              gridExtra_2.3              </span>
<span class="co">#&gt; [101] compiler_4.1.2              scater_1.23.7              </span>
<span class="co">#&gt; [103] tibble_3.1.6                KernSmooth_2.23-20         </span>
<span class="co">#&gt; [105] crayon_1.5.0                htmltools_0.5.2            </span>
<span class="co">#&gt; [107] mgcv_1.8-39                 later_1.3.0                </span>
<span class="co">#&gt; [109] tidyr_1.2.0                 DBI_1.1.2                  </span>
<span class="co">#&gt; [111] MASS_7.3-55                 Matrix_1.4-0               </span>
<span class="co">#&gt; [113] cli_3.2.0                   igraph_1.2.11              </span>
<span class="co">#&gt; [115] DR.SC_2.9                   GenomicRanges_1.46.1       </span>
<span class="co">#&gt; [117] pkgconfig_2.0.3             pkgdown_2.0.2              </span>
<span class="co">#&gt; [119] plotly_4.10.0               scuttle_1.4.0              </span>
<span class="co">#&gt; [121] spatstat.sparse_2.1-0       vipor_0.4.5                </span>
<span class="co">#&gt; [123] bslib_0.3.1                 XVector_0.34.0             </span>
<span class="co">#&gt; [125] CompQuadForm_1.4.3          stringr_1.4.0              </span>
<span class="co">#&gt; [127] digest_0.6.29               sctransform_0.3.3          </span>
<span class="co">#&gt; [129] RcppAnnoy_0.0.19            spatstat.data_2.1-2        </span>
<span class="co">#&gt; [131] rmarkdown_2.11              leiden_0.3.9               </span>
<span class="co">#&gt; [133] uwot_0.1.11                 DelayedMatrixStats_1.16.0  </span>
<span class="co">#&gt; [135] shiny_1.7.1                 lifecycle_1.0.1            </span>
<span class="co">#&gt; [137] nlme_3.1-155                jsonlite_1.8.0             </span>
<span class="co">#&gt; [139] BiocNeighbors_1.12.0        desc_1.4.0                 </span>
<span class="co">#&gt; [141] viridisLite_0.4.0           limma_3.50.1               </span>
<span class="co">#&gt; [143] fansi_1.0.2                 pillar_1.7.0               </span>
<span class="co">#&gt; [145] lattice_0.20-45             fastmap_1.1.0              </span>
<span class="co">#&gt; [147] httr_1.4.2                  survival_3.2-13            </span>
<span class="co">#&gt; [149] glue_1.6.2                  png_0.1-7                  </span>
<span class="co">#&gt; [151] stringi_1.7.6               sass_0.4.0                 </span>
<span class="co">#&gt; [153] textshaping_0.3.6           BiocSingular_1.10.0        </span>
<span class="co">#&gt; [155] memoise_2.0.1               irlba_2.3.5                </span>
<span class="co">#&gt; [157] future.apply_1.8.1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
