<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>PRECAST: simulation • PRECAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="PRECAST: simulation">
<meta property="og:description" content="PRECAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>PRECAST: simulation</h1>
                        <h4 data-toc-skip class="author">Wei Liu</h4>
            
            <h4 data-toc-skip class="date">2022-07-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/feiyoung/PRECAST/blob/HEAD/vignettes/PRECAST.Simu.Rmd" class="external-link"><code>vignettes/PRECAST.Simu.Rmd</code></a></small>
      <div class="hidden name"><code>PRECAST.Simu.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="generate-the-simulated-data">Generate the simulated data<a class="anchor" aria-label="anchor" href="#generate-the-simulated-data"></a>
</h2>
<p>First, we generate the multiple spatial transcriptomics data with lattice neighborhood, i.e. ST platform by using the function <code>gendata_seulist</code> in <code>PRECAST</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 4.1.3</span>
<span class="co">#&gt; Attaching SeuratObject</span>
<span class="co">#&gt; Attaching sp</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/feiyoung/PRECAST" class="external-link">PRECAST</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; Loading required package: gtools</span>
<span class="va">seuList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gendata_seulist.html">gendata_seulist</a></span><span class="op">(</span>height1<span class="op">=</span><span class="fl">20</span>, width1<span class="op">=</span><span class="fl">20</span>, height2<span class="op">=</span><span class="fl">24</span>, width2<span class="op">=</span><span class="fl">25</span>,p<span class="op">=</span><span class="fl">400</span>, K<span class="op">=</span><span class="fl">4</span>, alpha<span class="op">=</span><span class="fl">10</span>, sigma2<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>
<span class="va">seuList</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 400 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (400 features, 0 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 400 features across 600 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (400 features, 0 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> 
<span class="co">#&gt;          orig.ident nCount_RNA nFeature_RNA row col true_cluster</span>
<span class="co">#&gt; cell1_1       cell1       3024          225   1   1            2</span>
<span class="co">#&gt; cell1_2       cell1       2230          220   2   1            3</span>
<span class="co">#&gt; cell1_3       cell1       3981          222   3   1            1</span>
<span class="co">#&gt; cell1_4       cell1       2630          215   4   1            2</span>
<span class="co">#&gt; cell1_5       cell1       2367          231   5   1            2</span>
<span class="co">#&gt; cell1_6       cell1       4098          222   6   1            1</span>
<span class="co">#&gt; cell1_7       cell1       4212          222   7   1            1</span>
<span class="co">#&gt; cell1_8       cell1       3280          229   8   1            1</span>
<span class="co">#&gt; cell1_9       cell1       3011          216   9   1            1</span>
<span class="co">#&gt; cell1_10      cell1       2747          235  10   1            1</span>
<span class="co">## Must include the columns named "row" and "col" for saving the spatial coordinates</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="fit-precast-using-simulated-data">Fit PRECAST using simulated data<a class="anchor" aria-label="anchor" href="#fit-precast-using-simulated-data"></a>
</h2>
<div class="section level3">
<h3 id="prepare-the-precastobject-with-preprocessing-step-">Prepare the PRECASTObject with preprocessing step.<a class="anchor" aria-label="anchor" href="#prepare-the-precastobject-with-preprocessing-step-"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Create </span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/CreateiDRSCObject.html">CreatePRECASTObject</a></span><span class="op">(</span><span class="va">seuList</span><span class="op">)</span>
<span class="co">#&gt; Filter spots and features from Raw count data...</span>
<span class="co">#&gt;  </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Find the spatially variables genes by SPARK-X...</span>
<span class="co">#&gt; ## ===== SPARK-X INPUT INFORMATION ====</span>
<span class="co">#&gt; ## number of total samples: 400</span>
<span class="co">#&gt; ## number of total genes: 366</span>
<span class="co">#&gt; ## Running with single core, may take some time</span>
<span class="co">#&gt; ## Testing With Projection Kernel</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 1</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 2</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 3</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 4</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 5</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 1</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 2</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 3</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 4</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 5</span>
<span class="co">#&gt; Find the spatially variables genes by SPARK-X...</span>
<span class="co">#&gt; ## ===== SPARK-X INPUT INFORMATION ====</span>
<span class="co">#&gt; ## number of total samples: 600</span>
<span class="co">#&gt; ## number of total genes: 379</span>
<span class="co">#&gt; ## Running with single core, may take some time</span>
<span class="co">#&gt; ## Testing With Projection Kernel</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 1</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 2</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 3</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 4</span>
<span class="co">#&gt; ## Testing With Gaussian Kernel 5</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 1</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 2</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 3</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 4</span>
<span class="co">#&gt; ## Testing With Cosine Kernel 5</span>
<span class="co">#&gt; Warning in FUN(X[[i]], ...): Only 366 SVGs will be returned since the number of genes is less than 2000</span>
<span class="co">#&gt; Warning in FUN(X[[i]], ...): Only 379 SVGs will be returned since the number of genes is less than 2000</span>
<span class="co">#&gt; Warning in CreatePRECASTObject(seuList): Number of SVGs in one of sample is less</span>
<span class="co">#&gt; than 2000, so set minimum number of SVGs as gene.number=366</span>
<span class="co">#&gt; Select common SVGs  for multiple samples...</span>
<span class="co">#&gt; Filter spots and features from SVGs count data...</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-the-model-setting">Add the model setting<a class="anchor" aria-label="anchor" href="#add-the-model-setting"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## check the number of genes/features after filtering step</span>
<span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 348 features across 400 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (348 features, 348 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 348 features across 600 samples within 1 assay </span>
<span class="co">#&gt; Active assay: RNA (348 features, 348 variable features)</span>
<span class="co">#&gt;  1 dimensional reduction calculated: trueEmbed</span>

<span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/AddAdjList.html">AddAdjList</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, platform <span class="op">=</span> <span class="st">"ST"</span><span class="op">)</span>
<span class="co">#&gt; Neighbors were identified for 400 out of 400 spots.</span>
<span class="co">#&gt; Neighbors were identified for 600 out of 600 spots.</span>

<span class="co">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the information in the algorithm.</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddParSetting.html">AddParSetting</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, Sigma_equal<span class="op">=</span><span class="cn">FALSE</span>, maxIter<span class="op">=</span><span class="fl">30</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fit-precast">Fit PRECAST<a class="anchor" aria-label="anchor" href="#fit-precast"></a>
</h3>
<p>For function <code>PRECAST</code>, users can specify the number of clusters <span class="math inline">\(K\)</span> or set <code>K</code> to be an integer vector by using modified BIC(MBIC) to determine <span class="math inline">\(K\)</span>. First, we try using user-specified number of clusters. For convenience, we give a single K here.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co">### Given K</span></a>
<a class="sourceLine" id="cb4-2" title="2">PRECASTObj &lt;-<span class="st"> </span><span class="kw">PRECAST</span>(PRECASTObj, <span class="dt">K=</span><span class="dv">4</span>)</a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co">#&gt; Intergrative data info.: 2 samples, 348 genes X 1000 spots------</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">#&gt; PRECAST model setting: error_heter=TRUE, Sigma_equal=FALSE, Sigma_diag=TRUE, mix_prop_heter=TRUE</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co">#&gt; Start computing intial values...</span></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="co">#&gt; fitting ...</span></a>
<a class="sourceLine" id="cb4-7" title="7"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-8" title="8">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb4-9" title="9"><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</a>
<a class="sourceLine" id="cb4-10" title="10">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb4-11" title="11"><span class="st">  </span><span class="er">|===================================</span><span class="st">                                   </span><span class="er">|</span><span class="st">  </span><span class="dv">50</span>%</a>
<a class="sourceLine" id="cb4-12" title="12">  <span class="op">|</span><span class="st">                                                                            </span></a>
<a class="sourceLine" id="cb4-13" title="13"><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</a>
<a class="sourceLine" id="cb4-14" title="14"><span class="co">#&gt; ----Fitting PRECAST model----------------</span></a>
<a class="sourceLine" id="cb4-15" title="15"><span class="co">#&gt; variable initialize finish! </span></a>
<a class="sourceLine" id="cb4-16" title="16"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-17" title="17"><span class="co">#&gt; diff Energy = 0.103225 </span></a>
<a class="sourceLine" id="cb4-18" title="18"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-19" title="19"><span class="co">#&gt; iter = 2, loglik= 76227.562500, dloglik=1.000035 </span></a>
<a class="sourceLine" id="cb4-20" title="20"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-21" title="21"><span class="co">#&gt; diff Energy = 3.618338 </span></a>
<a class="sourceLine" id="cb4-22" title="22"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-23" title="23"><span class="co">#&gt; iter = 3, loglik= 83311.125000, dloglik=0.092927 </span></a>
<a class="sourceLine" id="cb4-24" title="24"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-25" title="25"><span class="co">#&gt; diff Energy = 0.753559 </span></a>
<a class="sourceLine" id="cb4-26" title="26"><span class="co">#&gt; diff Energy = 0.079758 </span></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="co">#&gt; iter = 4, loglik= 84979.523438, dloglik=0.020026 </span></a>
<a class="sourceLine" id="cb4-29" title="29"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-30" title="30"><span class="co">#&gt; diff Energy = 1.262075 </span></a>
<a class="sourceLine" id="cb4-31" title="31"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-32" title="32"><span class="co">#&gt; iter = 5, loglik= 85735.000000, dloglik=0.008890 </span></a>
<a class="sourceLine" id="cb4-33" title="33"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-34" title="34"><span class="co">#&gt; diff Energy = 3.770197 </span></a>
<a class="sourceLine" id="cb4-35" title="35"><span class="co">#&gt; diff Energy = 0.931205 </span></a>
<a class="sourceLine" id="cb4-36" title="36"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-37" title="37"><span class="co">#&gt; iter = 6, loglik= 86175.937500, dloglik=0.005143 </span></a>
<a class="sourceLine" id="cb4-38" title="38"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-39" title="39"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-40" title="40"><span class="co">#&gt; iter = 7, loglik= 86484.828125, dloglik=0.003584 </span></a>
<a class="sourceLine" id="cb4-41" title="41"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-42" title="42"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-43" title="43"><span class="co">#&gt; iter = 8, loglik= 86717.765625, dloglik=0.002693 </span></a>
<a class="sourceLine" id="cb4-44" title="44"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-45" title="45"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-46" title="46"><span class="co">#&gt; iter = 9, loglik= 86876.679688, dloglik=0.001833 </span></a>
<a class="sourceLine" id="cb4-47" title="47"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-48" title="48"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-49" title="49"><span class="co">#&gt; iter = 10, loglik= 87002.531250, dloglik=0.001449 </span></a>
<a class="sourceLine" id="cb4-50" title="50"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-51" title="51"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-52" title="52"><span class="co">#&gt; iter = 11, loglik= 87100.125000, dloglik=0.001122 </span></a>
<a class="sourceLine" id="cb4-53" title="53"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-54" title="54"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-55" title="55"><span class="co">#&gt; iter = 12, loglik= 87177.742188, dloglik=0.000891 </span></a>
<a class="sourceLine" id="cb4-56" title="56"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-57" title="57"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-58" title="58"><span class="co">#&gt; iter = 13, loglik= 87241.570312, dloglik=0.000732 </span></a>
<a class="sourceLine" id="cb4-59" title="59"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-60" title="60"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-61" title="61"><span class="co">#&gt; iter = 14, loglik= 87293.453125, dloglik=0.000595 </span></a>
<a class="sourceLine" id="cb4-62" title="62"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-63" title="63"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-64" title="64"><span class="co">#&gt; iter = 15, loglik= 87336.070312, dloglik=0.000488 </span></a>
<a class="sourceLine" id="cb4-65" title="65"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-66" title="66"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-67" title="67"><span class="co">#&gt; iter = 16, loglik= 87371.421875, dloglik=0.000405 </span></a>
<a class="sourceLine" id="cb4-68" title="68"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-69" title="69"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-70" title="70"><span class="co">#&gt; iter = 17, loglik= 87400.945312, dloglik=0.000338 </span></a>
<a class="sourceLine" id="cb4-71" title="71"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-72" title="72"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-73" title="73"><span class="co">#&gt; iter = 18, loglik= 87425.757812, dloglik=0.000284 </span></a>
<a class="sourceLine" id="cb4-74" title="74"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-75" title="75"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-76" title="76"><span class="co">#&gt; iter = 19, loglik= 87446.726562, dloglik=0.000240 </span></a>
<a class="sourceLine" id="cb4-77" title="77"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-78" title="78"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-79" title="79"><span class="co">#&gt; iter = 20, loglik= 87464.546875, dloglik=0.000204 </span></a>
<a class="sourceLine" id="cb4-80" title="80"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-81" title="81"><span class="co">#&gt; diff Energy = 0.741400 </span></a>
<a class="sourceLine" id="cb4-82" title="82"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-83" title="83"><span class="co">#&gt; iter = 21, loglik= 87479.015625, dloglik=0.000165 </span></a>
<a class="sourceLine" id="cb4-84" title="84"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-85" title="85"><span class="co">#&gt; diff Energy = 1.302642 </span></a>
<a class="sourceLine" id="cb4-86" title="86"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-87" title="87"><span class="co">#&gt; iter = 22, loglik= 87487.617188, dloglik=0.000098 </span></a>
<a class="sourceLine" id="cb4-88" title="88"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-89" title="89"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-90" title="90"><span class="co">#&gt; iter = 23, loglik= 87498.242188, dloglik=0.000121 </span></a>
<a class="sourceLine" id="cb4-91" title="91"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-92" title="92"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-93" title="93"><span class="co">#&gt; iter = 24, loglik= 87513.078125, dloglik=0.000170 </span></a>
<a class="sourceLine" id="cb4-94" title="94"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-95" title="95"><span class="co">#&gt; diff Energy = 0.704982 </span></a>
<a class="sourceLine" id="cb4-96" title="96"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-97" title="97"><span class="co">#&gt; iter = 25, loglik= 87521.445312, dloglik=0.000096 </span></a>
<a class="sourceLine" id="cb4-98" title="98"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-99" title="99"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-100" title="100"><span class="co">#&gt; iter = 26, loglik= 87535.031250, dloglik=0.000155 </span></a>
<a class="sourceLine" id="cb4-101" title="101"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-102" title="102"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-103" title="103"><span class="co">#&gt; iter = 27, loglik= 87548.132812, dloglik=0.000150 </span></a>
<a class="sourceLine" id="cb4-104" title="104"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-105" title="105"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-106" title="106"><span class="co">#&gt; iter = 28, loglik= 87553.265625, dloglik=0.000059 </span></a>
<a class="sourceLine" id="cb4-107" title="107"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-108" title="108"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-109" title="109"><span class="co">#&gt; iter = 29, loglik= 87558.023438, dloglik=0.000054 </span></a>
<a class="sourceLine" id="cb4-110" title="110"><span class="co">#&gt; predict Y and V! </span></a>
<a class="sourceLine" id="cb4-111" title="111"><span class="co">#&gt; Finish ICM step! </span></a>
<a class="sourceLine" id="cb4-112" title="112"><span class="co">#&gt; iter = 30, loglik= 87562.031250, dloglik=0.000046</span></a></code></pre></div>
<p>Select a best model and use ARI to check the performance of clustering</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## backup the fitting results in resList</span>
<span class="va">resList</span> <span class="op">&lt;-</span> <span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span>
<span class="co"># PRECASTObj@resList &lt;- resList</span>
<span class="va">PRECASTObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/selectModel.html">selectModel</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span>
<span class="va">true_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">seuList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">true_cluster</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span>
<span class="co">#&gt; List of 2</span>
<span class="co">#&gt;  $ : Named num [1:400] 2 3 1 2 2 1 1 1 1 1 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:400] "cell1_1" "cell1_2" "cell1_3" "cell1_4" ...</span>
<span class="co">#&gt;  $ : Named num [1:600] 3 3 3 1 1 1 1 1 1 1 ...</span>
<span class="co">#&gt;   ..- attr(*, "names")= chr [1:600] "cell2_1" "cell2_2" "cell2_3" "cell2_4" ...</span>
<span class="fu">mclust</span><span class="fu">::</span><span class="fu"><a href="https://mclust-org.github.io/mclust/reference/adjustedRandIndex.html" class="external-link">adjustedRandIndex</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">resList</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">true_cluster</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.932775</span></code></pre></div>
<p>Integrate the two samples by the function <code>IntegrateSpaData</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IntegrateSpaData.html">IntegrateSpaData</a></span><span class="op">(</span><span class="va">PRECASTObj</span>, species<span class="op">=</span><span class="st">'unknown'</span><span class="op">)</span>
<span class="co">#&gt; Using only PRECAST results to obtain the batch corrected gene expressions</span>
<span class="co">#&gt;  since species is unkown or the genelist in PRECASTObj has less than 5 overlapp </span>
<span class="co">#&gt;  with the housekeeping genes of given species.</span>
<span class="co">#&gt; Users can specify the custom_housekeep by themselves to use the housekeeping genes based methods.</span>
<span class="va">seuInt</span> 
<span class="co">#&gt; An object of class Seurat </span>
<span class="co">#&gt; 348 features across 1000 samples within 1 assay </span>
<span class="co">#&gt; Active assay: PRE_CAST (348 features, 0 variable features)</span>
<span class="co">#&gt;  2 dimensional reductions calculated: PRECAST, position</span>
<span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p12</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>,point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">p12</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># users can plot each sample by setting combine=FALSE</span></code></pre></div>
<p>Show the spatial UMAP/tNSE RGB plot</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddUMAP.html">AddUMAP</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span> 
<span class="fu"><a href="../reference/SpaPlot.html">SpaPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, batch<span class="op">=</span><span class="cn">NULL</span>,item<span class="op">=</span><span class="st">'RGB_UMAP'</span>,point_size<span class="op">=</span><span class="fl">2</span>, combine<span class="op">=</span><span class="cn">TRUE</span>, text_size<span class="op">=</span><span class="fl">15</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#seuInt &lt;- AddTSNE(seuInt) </span>
<span class="co">#SpaPlot(seuInt, batch=NULL,item='RGB_TSNE',point_size=2, combine=T, text_size=15)</span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to check the performance of integration.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AddTSNE.html">AddTSNE</a></span><span class="op">(</span><span class="va">seuInt</span>, n_comp <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> 
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span>
<span class="va">cols_cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#E04D50"</span>, <span class="st">"#4374A5"</span>, <span class="st">"#F08A21"</span>,<span class="st">"#2AB673"</span>, <span class="st">"#FCDDDE"</span>,  <span class="st">"#70B5B0"</span>, <span class="st">"#DFE0EE"</span> ,<span class="st">"#D0B14C"</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>,  font_family<span class="op">=</span><span class="st">'serif'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span><span class="op">)</span> <span class="co"># Times New Roman</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, item<span class="op">=</span><span class="st">'batch'</span>, point_size <span class="op">=</span> <span class="fl">1</span>,  font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span>
<span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> </code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># It is noted that only sample batch 1 has cluster 4, and only sample batch 2 has cluster 7. </span></code></pre></div>
<p>Show the UMAP plot based on the extracted features from PRECAST.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/dimPlot.html">dimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'UMAP3'</span>, item<span class="op">=</span><span class="st">'cluster'</span>, cols<span class="op">=</span><span class="va">cols_cluster</span>, font_family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Users can also use the visualization functions in Seurat package:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'position'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'tSNE'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-2.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seuInt</span>, reduction <span class="op">=</span> <span class="st">'PRECAST'</span><span class="op">)</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-12-3.png" width="700"></p>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dat_deg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Calculating cluster 1</span>
<span class="co">#&gt; Calculating cluster 2</span>
<span class="co">#&gt; Calculating cluster 3</span>
<span class="co">#&gt; Calculating cluster 4</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'dplyr' was built under R version 4.1.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">10</span>
<span class="va">dat_deg</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html" class="external-link">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">top10</span>

<span class="va">seuInt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seuInt</span><span class="op">)</span>
<span class="co">#&gt; Centering and scaling data matrix</span>
<span class="va">seus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">seuInt</span>, downsample <span class="op">=</span> <span class="fl">400</span><span class="op">)</span>
<span class="va">color_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">seus</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="co">#&gt; Warning: package 'ggplot2' was built under R version 4.1.3</span>

<span class="co">## HeatMap</span>
<span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/doHeatmap.html">doHeatmap</a></span><span class="op">(</span><span class="va">seus</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span>, cell_label<span class="op">=</span> <span class="st">"Domain"</span>,
                grp_label <span class="op">=</span> <span class="cn">F</span>, grp_color <span class="op">=</span> <span class="va">cols_cluster</span>,
                pt_size<span class="op">=</span><span class="fl">6</span>,slot <span class="op">=</span> <span class="st">'scale.data'</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">16</span><span class="op">)</span>,
        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">18</span>, face<span class="op">=</span><span class="st">'bold'</span><span class="op">)</span>,
        axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">7</span>, face<span class="op">=</span> <span class="st">"italic"</span>, family<span class="op">=</span><span class="st">'serif'</span><span class="op">)</span><span class="op">)</span>
<span class="va">p1</span></code></pre></div>
<p><img src="PRECAST.Simu_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.1.2 (2021-11-01)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows 10 x64 (build 22000)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=English_Singapore.1252  LC_CTYPE=English_Singapore.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=English_Singapore.1252 LC_NUMERIC=C                      </span>
<span class="co">#&gt; [5] LC_TIME=English_Singapore.1252    </span>
<span class="co">#&gt; system code page: 936</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ggplot2_3.3.6      dplyr_1.0.9        patchwork_1.1.1    PRECAST_1.2       </span>
<span class="co">#&gt; [5] gtools_3.9.2.2     sp_1.5-0           SeuratObject_4.1.0 Seurat_4.1.1      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] utf8_1.2.2                  reticulate_1.25            </span>
<span class="co">#&gt;   [3] tidyselect_1.1.2            htmlwidgets_1.5.4          </span>
<span class="co">#&gt;   [5] grid_4.1.2                  BiocParallel_1.28.3        </span>
<span class="co">#&gt;   [7] Rtsne_0.16                  munsell_0.5.0              </span>
<span class="co">#&gt;   [9] ScaledMatrix_1.2.0          codetools_0.2-18           </span>
<span class="co">#&gt;  [11] ragg_1.2.2                  ica_1.0-2                  </span>
<span class="co">#&gt;  [13] future_1.26.1               miniUI_0.1.1.1             </span>
<span class="co">#&gt;  [15] withr_2.5.0                 spatstat.random_2.2-0      </span>
<span class="co">#&gt;  [17] colorspace_2.0-3            progressr_0.10.1           </span>
<span class="co">#&gt;  [19] Biobase_2.54.0              highr_0.9                  </span>
<span class="co">#&gt;  [21] knitr_1.37                  rstudioapi_0.13            </span>
<span class="co">#&gt;  [23] stats4_4.1.2                SingleCellExperiment_1.16.0</span>
<span class="co">#&gt;  [25] ROCR_1.0-11                 tensor_1.5                 </span>
<span class="co">#&gt;  [27] listenv_0.8.0               labeling_0.4.2             </span>
<span class="co">#&gt;  [29] MatrixGenerics_1.6.0        GenomeInfoDbData_1.2.7     </span>
<span class="co">#&gt;  [31] polyclip_1.10-0             farver_2.1.0               </span>
<span class="co">#&gt;  [33] rprojroot_2.0.3             parallelly_1.32.0          </span>
<span class="co">#&gt;  [35] vctrs_0.4.1                 generics_0.1.2             </span>
<span class="co">#&gt;  [37] xfun_0.29                   ggthemes_4.2.4             </span>
<span class="co">#&gt;  [39] R6_2.5.1                    GenomeInfoDb_1.30.1        </span>
<span class="co">#&gt;  [41] ggbeeswarm_0.6.0            rsvd_1.0.5                 </span>
<span class="co">#&gt;  [43] bitops_1.0-7                spatstat.utils_2.3-1       </span>
<span class="co">#&gt;  [45] cachem_1.0.6                DelayedArray_0.20.0        </span>
<span class="co">#&gt;  [47] assertthat_0.2.1            promises_1.2.0.1           </span>
<span class="co">#&gt;  [49] scales_1.2.0                rgeos_0.5-9                </span>
<span class="co">#&gt;  [51] beeswarm_0.4.0              gtable_0.3.0               </span>
<span class="co">#&gt;  [53] beachmat_2.10.0             globals_0.15.0             </span>
<span class="co">#&gt;  [55] goftest_1.2-3               rlang_1.0.2                </span>
<span class="co">#&gt;  [57] systemfonts_1.0.4           splines_4.1.2              </span>
<span class="co">#&gt;  [59] lazyeval_0.2.2              spatstat.geom_2.4-0        </span>
<span class="co">#&gt;  [61] yaml_2.3.5                  reshape2_1.4.4             </span>
<span class="co">#&gt;  [63] abind_1.4-5                 httpuv_1.6.5               </span>
<span class="co">#&gt;  [65] tools_4.1.2                 ellipsis_0.3.2             </span>
<span class="co">#&gt;  [67] spatstat.core_2.4-4         jquerylib_0.1.4            </span>
<span class="co">#&gt;  [69] RColorBrewer_1.1-3          BiocGenerics_0.40.0        </span>
<span class="co">#&gt;  [71] ggridges_0.5.3              Rcpp_1.0.8.3               </span>
<span class="co">#&gt;  [73] plyr_1.8.7                  sparseMatrixStats_1.6.0    </span>
<span class="co">#&gt;  [75] zlibbioc_1.40.0             purrr_0.3.4                </span>
<span class="co">#&gt;  [77] RCurl_1.98-1.6              rpart_4.1.16               </span>
<span class="co">#&gt;  [79] deldir_1.0-6                viridis_0.6.2              </span>
<span class="co">#&gt;  [81] pbapply_1.5-0               cowplot_1.1.1              </span>
<span class="co">#&gt;  [83] S4Vectors_0.32.3            zoo_1.8-10                 </span>
<span class="co">#&gt;  [85] SummarizedExperiment_1.24.0 ggrepel_0.9.1              </span>
<span class="co">#&gt;  [87] cluster_2.1.2               fs_1.5.2                   </span>
<span class="co">#&gt;  [89] magrittr_2.0.3              GiRaF_1.0.1                </span>
<span class="co">#&gt;  [91] data.table_1.14.2           scattermore_0.8            </span>
<span class="co">#&gt;  [93] lmtest_0.9-40               RANN_2.6.1                 </span>
<span class="co">#&gt;  [95] fitdistrplus_1.1-8          matrixStats_0.62.0         </span>
<span class="co">#&gt;  [97] mime_0.12                   evaluate_0.15              </span>
<span class="co">#&gt;  [99] xtable_1.8-4                mclust_5.4.10              </span>
<span class="co">#&gt; [101] IRanges_2.28.0              gridExtra_2.3              </span>
<span class="co">#&gt; [103] compiler_4.1.2              scater_1.25.1              </span>
<span class="co">#&gt; [105] tibble_3.1.7                KernSmooth_2.23-20         </span>
<span class="co">#&gt; [107] crayon_1.5.1                htmltools_0.5.2            </span>
<span class="co">#&gt; [109] mgcv_1.8-39                 later_1.3.0                </span>
<span class="co">#&gt; [111] tidyr_1.2.0                 DBI_1.1.2                  </span>
<span class="co">#&gt; [113] MASS_7.3-55                 Matrix_1.4-0               </span>
<span class="co">#&gt; [115] cli_3.2.0                   igraph_1.3.2               </span>
<span class="co">#&gt; [117] DR.SC_2.9                   GenomicRanges_1.46.1       </span>
<span class="co">#&gt; [119] pkgconfig_2.0.3             pkgdown_2.0.2              </span>
<span class="co">#&gt; [121] plotly_4.10.0               scuttle_1.4.0              </span>
<span class="co">#&gt; [123] spatstat.sparse_2.1-1       vipor_0.4.5                </span>
<span class="co">#&gt; [125] bslib_0.3.1                 XVector_0.34.0             </span>
<span class="co">#&gt; [127] CompQuadForm_1.4.3          stringr_1.4.0              </span>
<span class="co">#&gt; [129] digest_0.6.29               sctransform_0.3.3          </span>
<span class="co">#&gt; [131] RcppAnnoy_0.0.19            spatstat.data_2.2-0        </span>
<span class="co">#&gt; [133] rmarkdown_2.11              leiden_0.4.2               </span>
<span class="co">#&gt; [135] uwot_0.1.11                 DelayedMatrixStats_1.16.0  </span>
<span class="co">#&gt; [137] shiny_1.7.1                 lifecycle_1.0.1            </span>
<span class="co">#&gt; [139] nlme_3.1-155                jsonlite_1.8.0             </span>
<span class="co">#&gt; [141] BiocNeighbors_1.12.0        limma_3.50.1               </span>
<span class="co">#&gt; [143] desc_1.4.0                  viridisLite_0.4.0          </span>
<span class="co">#&gt; [145] fansi_1.0.3                 pillar_1.7.0               </span>
<span class="co">#&gt; [147] lattice_0.20-45             fastmap_1.1.0              </span>
<span class="co">#&gt; [149] httr_1.4.3                  survival_3.2-13            </span>
<span class="co">#&gt; [151] glue_1.6.2                  png_0.1-7                  </span>
<span class="co">#&gt; [153] stringi_1.7.6               sass_0.4.1                 </span>
<span class="co">#&gt; [155] textshaping_0.3.6           BiocSingular_1.10.0        </span>
<span class="co">#&gt; [157] memoise_2.0.1               irlba_2.3.5                </span>
<span class="co">#&gt; [159] future.apply_1.9.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
