<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ICM-EM algorithm implementation — ICM.EM • PRECAST</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ICM-EM algorithm implementation — ICM.EM"><meta property="og:description" content="ICM-EM algorithm for fitting PRECAST model"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PRECAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/PRECAST.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/PRECAST.BreastCancer.html">PRECAST: Human Breast Cancer Data Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.DLPFC.html">PRECAST: DLPFC Single Sample Analysis</a>
    </li>
    <li>
      <a href="../articles/PRECAST.Simu.html">PRECAST: simulation</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/feiyoung/PRECAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ICM-EM algorithm implementation</h1>
    
    <div class="hidden name"><code>idrsc.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>ICM-EM algorithm for fitting PRECAST model</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">ICM.EM</span><span class="op">(</span><span class="va">XList</span>, <span class="va">q</span>, <span class="va">K</span>, AdjList<span class="op">=</span><span class="cn">NULL</span>,  Adjlist_car<span class="op">=</span><span class="cn">NULL</span>, posList <span class="op">=</span> <span class="cn">NULL</span>, 
      platform <span class="op">=</span> <span class="st">"ST"</span>, beta_grid<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.2</span>,<span class="fl">4</span>, by<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>,maxIter_ICM<span class="op">=</span><span class="fl">6</span>,
      maxIter<span class="op">=</span><span class="fl">20</span>, epsLogLik<span class="op">=</span><span class="fl">1e-5</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>,mix_prop_heter<span class="op">=</span><span class="cn">TRUE</span>, 
      Sigma_equal<span class="op">=</span><span class="cn">FALSE</span>, Sigma_diag<span class="op">=</span><span class="cn">TRUE</span>,error_heter<span class="op">=</span><span class="cn">TRUE</span>, Sp2<span class="op">=</span><span class="cn">TRUE</span>,
      wpca_int<span class="op">=</span><span class="cn">FALSE</span>, int.model<span class="op">=</span><span class="st">'EEE'</span>, seed<span class="op">=</span><span class="fl">1</span>,coreNum <span class="op">=</span> <span class="fl">1</span>, coreNum_int<span class="op">=</span><span class="va">coreNum</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>XList</dt>
<dd><p>an M-length list consisting of multiple matrices with class <code>dgCMatrix</code> or <code>matrix</code> that specify the log-normalization gene expression matrix for each data sample used for iDR-SC model.</p></dd>
<dt>q</dt>
<dd><p>a positive integer, specify the number of latent features to be extracted, default as 15.</p></dd>
<dt>K</dt>
<dd><p>a positive  integer allowing scalar or vector, specify the number of clusters in model fitting.</p></dd>
<dt>AdjList</dt>
<dd><p>an M-length list of sparse matrices with class <code>dgCMatrix</code>, specify the adjacency matrix used for Potts model in iDR-SC. We provide this interface for those users  who would like to define the adjacency matrix by their own.</p></dd>
<dt>Adjlist_car</dt>
<dd><p>an M-length list of sparse matrices with class <code>dgCMatrix</code>, specify the adjacency matrix used for CAR model in iDR-SC, default as AdjList in the Potts model. We provide this interface for those users  who would like to use the different adjacency matrix in CAR model.</p></dd>
<dt>posList</dt>
<dd><p>an M-length list composed by spatial coordinate matrix for each data sample.</p></dd>
<dt>platform</dt>
<dd><p>a string, specify the platform of the provided data, default as "Visium".
  There are many platforms to be supported, including ("Visuim", "ST", "SeqFISH", 'merFISH', 'slide-seqv2', 'seqscope', "HDST"). If AdjList is not given, the The platform helps to calculate the adjacency matrix by defining the neighbors.</p></dd>
<dt>beta_grid</dt>
<dd><p>an optional vector of positive value, the candidate set of the smoothing parameter to be searched by the grid-search optimization approach.</p></dd>
<dt>maxIter_ICM</dt>
<dd><p>an optional positive value, represents the maximum iterations of ICM.</p></dd>
<dt>maxIter</dt>
<dd><p>an optional positive value, represents the maximum iterations of EM.</p></dd>
<dt>epsLogLik</dt>
<dd><p>an optional positive vlaue, tolerance vlaue of relative variation rate of the observed pseudo log-loglikelihood value, defualt as '1e-5'.</p></dd>
<dt>verbose</dt>
<dd><p>an optional logical value, whether output the information of the ICM-EM algorithm.</p></dd>
<dt>mix_prop_heter</dt>
<dd><p>an optional logical value, specify whether betar are distict, default as <code>TRUE</code>.</p></dd>
<dt>Sigma_equal</dt>
<dd><p>an optional logical value, specify whether Sigmaks are equal, default as FALSE.</p></dd>
<dt>Sigma_diag</dt>
<dd><p>an optional logical value, specify whether Sigmaks are diagonal matrices, default as <code>TRUE</code>.</p></dd>
<dt>error_heter</dt>
<dd><p>an optional logical value, whether use the heterogenous error for DR-SC model, default as <code>TRUE</code>. If <code>error_heter=FALSE</code>, then the homogenuous error is used for probabilistic PCA model in iDR-SC.</p></dd>
<dt>Sp2</dt>
<dd><p>an optional logical value, whether add the ICAR model component in the model, default as TRUE. We provide this interface for those users  who don't want to include the  ICAR model.</p></dd>
<dt>wpca_int</dt>
<dd><p>an optional logical value, means whether use the weighted PCA to obtain the initial values of loadings and other paramters, default as <code>FALSE</code> which means the ordinary PCA  is used.</p></dd>
<dt>int.model</dt>
<dd><p>an optional string, specify which Gaussian mixture model is used in evaluting the initial values for PRECAST, default as "EEE"; and see <code>Mclust</code> for more models' names.</p></dd>
<dt>seed</dt>
<dd><p>an optional integer, the random seed in fitting PRECAST model.</p></dd>
<dt>coreNum</dt>
<dd><p>an optional positive integer, means the number of thread used in parallel computating.</p></dd>
<dt>coreNum_int</dt>
<dd><p>an optional positive integer, means the number of cores used in parallel computation for initial values when <code>K</code> is a vector, default as same as <code>coreNum</code>.</p></dd>
</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Nothing</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    <p>ICM.EM returns a <a href="https://rdrr.io/r/base/list.html" class="external-link">list</a> with class "SeqKiDRSC_Object" with the number of components equal to the length of <code>K</code>, where each component includes the model fitting results for one number of cluster and is a list consisting of following components:</p>
<dl><dt>cluster</dt>
<dd><p>an M-length list that includes the inferred class labels for each data sample.</p></dd>
<dt>hZ</dt>
<dd><p>an M-length list that includes the batch corrected low-dimensional embeddings for each data sample.</p></dd>
<dt>hV</dt>
<dd><p>an M-length list that includes the estimate the ICAR component for each sample.</p></dd>
<dt>Rf</dt>
<dd><p>an M-length list that includes the posterior probability of domain clusters for each sample.</p></dd>
<dt>beta</dt>
<dd><p>an M-length vector that includes the estimated smoothing parameters for each sample.</p></dd>
<dt>Mu</dt>
<dd><p>mean vectors of mixtures components.</p></dd>
<dt>Sigma</dt>
<dd><p>covariance matrix of mixtures components.</p></dd>
<dt>W</dt>
<dd><p>estimated loading matrix</p></dd>
<dt>Lam</dt>
<dd><p>estimated variance of errors in probabilistic PCA model</p></dd>
<dt>loglik</dt>
<dd><p>pseudo observed log-likelihood.</p></dd>
</dl></div>
    <div id="references">
    <h2>References</h2>
    <p><a href="https://www.biorxiv.org/content/10.1101/2021.12.25.474153v4" class="external-link">Wei Liu, Xu Liao, Yi Yang, Huazhen Lin, Joe Yeong, Xiang Zhou, Xingjie Shi and Jin Liu. (2022) Joint dimension reduction and clustering analysis for single-cell RNA-seq and spatial transcriptomics data</a></p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Wei Liu</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>nothing</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>None</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># \donttest{</span></span>
<span class="r-in">  <span class="co">## we generate the spatial transcriptomics data with lattice neighborhood, i.e. ST platform.</span></span>
<span class="r-in">  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span class="r-in">  <span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">10</span>; <span class="va">K</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span class="r-in">  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">posList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">row</span>, <span class="va">x</span><span class="op">$</span><span class="va">col</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">AdjList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">posList</span>, <span class="va">getAdj_reg</span>, platform<span class="op">=</span><span class="st">'ST'</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Neighbors were identified for 900 out of 900 spots.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Neighbors were identified for 900 out of 900 spots.</span>
<span class="r-in">  <span class="va">XList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">PRECASTObj</span><span class="op">@</span><span class="va">seulist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">  <span class="va">XList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">XList</span>, <span class="va">scale</span>, scale<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span class="r-in">  <span class="co">## For illustration, maxIter is set to 4</span></span>
<span class="r-in">  <span class="va">resList</span> <span class="op">&lt;-</span> <span class="fu">ICM.EM</span><span class="op">(</span><span class="va">XList</span>,AdjList <span class="op">=</span> <span class="va">AdjList</span>, maxIter<span class="op">=</span><span class="fl">4</span>,</span>
<span class="r-in">                   q<span class="op">=</span><span class="va">q</span>, K<span class="op">=</span><span class="va">K</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Intergrative data info.: 2 samples, 97 genes X 1800 spots------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PRECAST model setting: error_heter=TRUE, Sigma_equal=FALSE, Sigma_diag=TRUE, mix_prop_heter=TRUE</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Start computing intial values... </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> fitting ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |===================================                                   |  50%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ----Fitting PRECAST model----------------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> variable initialize finish! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> predict Y and V! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> diff Energy = 44.205532 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Finish ICM step! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 2, loglik= -59510.035156, dloglik=0.999972 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> predict Y and V! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> diff Energy = 34.843510 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> diff Energy = 4.917973 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Finish ICM step! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 3, loglik= -52992.929688, dloglik=0.109513 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> predict Y and V! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> diff Energy = 55.378534 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> diff Energy = 3.051830 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Finish ICM step! </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> iter = 4, loglik= -50844.128906, dloglik=0.040549 </span>
<span class="r-in"><span class="co"># }</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Wei Liu, Yi Yang, Jin Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

